### PR TITLE
[amberscript] Add REPEAT command

### LINKED ISSUE #425: Add REPEAT to AmberScript
We want the abilty to run a given draw command multiple times. To that end, add:

```
REPEAT n
  ...
END
```

### PR DESCRIPTION
**Pull request by dj2**
This CL adds the REPEAT directive to AmberScript. This allows running a
given set of commands `n` times.

Fixes #425.

### CODE REVIEW

File: docs/amber_script.md [hunk 1]
```diff
@@ -383,7 +401,7 @@ void main() {
 }
 END  # shader
 
-BUFFER kComputeBuffer TYPE vec2<int32> SIZE 524288 FILL 0
+BUFFER kComputeBuffer DATA_TYPE vec2<int32> SIZE 524288 FILL 0
```
Comment (dneto0) [inline docs/amber_script.md]: This is unrelated to the REPEAT functionality.

And I think ELEM_TYPE is better than DATA_TYPE.  Because actually the DATA_TYPE is _array of_ ....

File: src/amberscript/parser.cc [hunk 2]
```diff
@@ -96,6 +90,7 @@ Result Parser::Parse(const std::string& data) {
     if (!r.IsSuccess())
       return Result(make_error(r.Error()));
   }
+  script_->SetCommands(std::move(command_list_));
```
Comment (dneto0) [inline src/amberscript/parser.cc]: Maybe it's a subtle point, but this actually does ensure that command_list is empty afterward.
The std::vector move constructor guarantees it.

File: src/amberscript/parser.cc [hunk 3]
```diff
@@ -1309,5 +1326,50 @@ Result Parser::ParseDeviceFeature() {
   return ValidateEndOfStatement("DEVICE_FEATURE command");
 }
 
+Result Parser::ParseRepeat() {
```
Comment (dneto0) [inline src/amberscript/parser.cc]: This whole thing works without a stack because REPEAT is not itself repeatable.  Works for me.

File: src/amberscript/parser.h [hunk 4]
```diff
@@ -67,6 +67,9 @@ class Parser : public amber::Parser {
   Result ParseExpect();
   Result ParseCopy();
   Result ParseDeviceFeature();
+  Result ParseRepeat();
+  bool isRepeatable(const std::string& name) const;
```
Comment (dneto0) [inline src/amberscript/parser.h]: nit: Is there a reason this is lower-case-first?

File: src/vkscript/command_parser.cc [hunk 5]
```diff
@@ -647,7 +647,13 @@ Result CommandParser::ProcessSSBO() {
       return Result("Invalid size value for ssbo command: " +
                     token->ToOriginalString());
 
-    cmd->SetSize(token->AsUint32());
+    uint32_t size = token->AsUint32();
+    cmd->SetSize(size);
+
+    // Resize the buffer so we'll correctly create the descriptor sets.
+    auto* buf = cmd->GetBuffer();
```
Comment (dneto0) [inline src/vkscript/command_parser.cc]: Is this actually related?

File: src/vulkan/engine_vulkan.cc [hunk 6]
```diff
@@ -211,6 +211,29 @@ Result EngineVulkan::CreatePipeline(amber::Pipeline* pipeline) {
     info.vk_pipeline->AsGraphics()->SetIndexBuffer(buf);
   }
 
+  for (const auto& buf_info : pipeline->GetBuffers()) {
```
Comment (dneto0) [inline src/vulkan/engine_vulkan.cc]: Is this actually related?

File: src/vulkan/pipeline.cc [hunk 7]
```diff
@@ -244,6 +244,8 @@ Result Pipeline::AddPushConstant(const BufferCommand* command) {
 Result Pipeline::AddDescriptor(const BufferCommand* cmd) {
   if (cmd == nullptr)
     return Result("Pipeline::AddDescriptor BufferCommand is nullptr");
+  if (cmd->IsPushConstant())
```
Comment (dneto0) [inline src/vulkan/pipeline.cc]: is this  related?

File: src/vulkan/pipeline.cc [hunk 8]
```diff
@@ -296,12 +298,14 @@ Result Pipeline::AddDescriptor(const BufferCommand* cmd) {
         "and binding");
   }
 
-  auto* buf_desc = static_cast<BufferDescriptor*>(desc);
-  Result r =
-      buf_desc->AddToBuffer(cmd->GetDatumType().GetType(), cmd->GetOffset(),
-                            cmd->GetSize(), cmd->GetValues());
-  if (!r.IsSuccess())
-    return r;
+  if (!cmd->GetValues().empty()) {
```
Comment (dneto0) [inline src/vulkan/pipeline.cc]: Is this related?

File: docs/amber_script.md (hunk 1 above)
Comment (dj2) [reply docs/amber_script.md]: The field is already called DATA_TYPE, this just fixes the examples to match the parser.

File: src/amberscript/parser.cc (hunk 2 above)
Comment (dj2) [reply src/amberscript/parser.cc]: Yes, but I need to swap it back to get the previous command list entries back into the list.

File: src/amberscript/parser.h (hunk 4 above)
Comment (dj2) [reply src/amberscript/parser.h]: fix

File: src/vkscript/command_parser.cc (hunk 5 above)
Comment (dj2) [reply src/vkscript/command_parser.cc]: I order to make the repeat.amber script work but buffers needed to work. To make buffers work, requires the ssbo's to be sized correctly. So ... kinda?

File: src/vulkan/engine_vulkan.cc (hunk 6 above)
Comment (dj2) [reply src/vulkan/engine_vulkan.cc]: ditto

File: src/vulkan/pipeline.cc (hunk 7 above)
Comment (dj2) [reply src/vulkan/pipeline.cc]: ditto

File: src/vulkan/pipeline.cc (hunk 8 above)
Comment (dj2) [reply src/vulkan/pipeline.cc]: ditto
